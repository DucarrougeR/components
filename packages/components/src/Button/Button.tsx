import omit from 'lodash/omit'
import { rem, rgba } from 'polished'
import React, { FunctionComponent, Ref } from 'react'
import styled, { css, StyledComponent, withTheme } from 'styled-components'
import { merge } from 'styled-system'
import { IconNames } from '../../icons/build/IconNames'
import {
  CustomizableAttributes,
  RampSizes,
  SemanticColor,
  SemanticColors,
  SizeLarge,
  SizeMedium,
  SizeSmall,
  SizeXLarge,
  SizeXSmall,
  ThemedProps,
} from '@looker/design-tokens'

import { Box, BoxPropsWithout } from '../Box'
import { Icon } from '../Icon/Icon'

export type ButtonSizes = SizeXSmall | SizeSmall | SizeMedium | SizeLarge
export type ButtonSpacingSizes =
  | SizeXSmall
  | SizeSmall
  | SizeMedium
  | SizeLarge
  | SizeXLarge

export interface ButtonProps
  extends BoxPropsWithout<HTMLButtonElement, 'color' | 'size'> {
  /**
   * Defines the color of the button. Can be the string name of a color listed in the color theme, or a color object.
   * @default "primary"
   */
  color?: keyof SemanticColors | SemanticColor
  /**
   * Determines if the button is disabled.
   * @default false
   */
  disabled?: boolean
  /**
   * Defines the size of the button.
   * @default "medium"
   */
  size?: ButtonSizes
  /**
   * The type of button to define
   * @default "button"
   */
  type?: 'submit' | 'reset' | 'button' | 'menu'
  /**
   * Defines the variant or mapping of colors to style properties, like border of the button.
   * @default "default"
   */
  variant?: 'default' | 'outline' | 'transparent'
  iconBefore?: IconNames | undefined
  iconAfter?: IconNames | undefined
}

type ComponentType = FunctionComponent<ThemedProps<ButtonProps>>
type StyledComponentType = StyledComponent<
  ComponentType,
  ThemedProps<ButtonProps>
>

const variantCommonProps = (color: SemanticColor) => {
  return {
    borderStyle: 'solid',
    borderWidth: rem(1),
    // tslint:disable-next-line:object-literal-sort-keys
    '&:focus': {
      boxShadow: `0 0 0 0.15rem ${rgba(color.main, 0.25)}`,
    },
    '&[disabled]': {
      cursor: 'default',
      filter: 'grayscale(0.3)',
      opacity: '0.25',
    },
  }
}

const defaultVariant = (color: SemanticColor) => {
  return merge(variantCommonProps(color), {
    background: color.main,
    borderColor: color.main,
    color: color.text,
    // tslint:disable-next-line:object-literal-sort-keys
    '&:hover, &:focus, &.hover': {
      background: color.dark,
      borderColor: color.dark,
    },
    '&:active, &.active': {
      background: color.darker,
      borderColor: color.darker,
    },
    '&[disabled]': {
      '&:hover, &:active, &:focus': {
        backgroundColor: color.main,
        borderColor: color.borderColor,
      },
    },
  })
}

const outlineVariant = (
  color: SemanticColor,
  props: ThemedProps<ButtonProps>
) => {
  return merge(variantCommonProps(color), {
    background: props.theme.colors.palette.white,
    borderColor: color.borderColor,
    color: color.main,
    // tslint:disable-next-line:object-literal-sort-keys
    '&:hover, &:focus, &.hover': {
      background: props.theme.colors.palette.white,
      borderColor: color.main,
      color: color.darker,
    },
    '&:active, &.active': {
      background: color.main,
      borderColor: color.main,
      color: color.text,
    },
    '&[disabled]': {
      '&:hover, &:active, &:focus': {
        backgroundColor: props.theme.colors.palette.white,
        borderColor: color.borderColor,
        color: color.main,
      },
    },
  })
}

const transparentVariant = (
  color: SemanticColor,
  props: ThemedProps<ButtonProps>
) => {
  return merge(variantCommonProps(color), {
    background: props.theme.colors.palette.transparent,
    borderColor: props.theme.colors.palette.transparent,
    color: color.main,
    // tslint:disable-next-line:object-literal-sort-keys
    '&:hover, &:focus, &.hover': {
      background: color.lighter,
      borderColor: color.lighter,
      color: color.altText,
    },
    '&:active, &.active': {
      background: color.light,
      borderColor: color.light,
      color: color.altText,
    },
    '&[disabled]': {
      '&:hover, &:active, &:focus': {
        backgroundColor: props.theme.colors.palette.transparent,
        borderColor: props.theme.colors.palette.transparent,
        color: color.main,
      },
    },
  })
}

const variantHelper = (props: ThemedProps<ButtonProps>) => {
  let color
  if (typeof props.color === 'string' || props.color === undefined) {
    color = props.theme.colors.semanticColors[props.color || 'primary']
  } else {
    color = props.color
  }
  switch (props.variant || 'default') {
    case 'transparent':
      return transparentVariant(color, props)
    case 'outline':
      return outlineVariant(color, props)
    case 'default':
    default:
      return defaultVariant(color)
  }
}

// Accepts a number and subtracts two from it to accout for border in button height
function calcLineHeight(size: number) {
  return `${size - 2}px`
}

function sizeHelper(props: ThemedProps<ButtonProps>) {
  const sizes: Record<ButtonSizes, [RampSizes, string, ButtonSpacingSizes]> = {
    large: ['xlarge', calcLineHeight(44), 'large'],
    medium: ['medium', calcLineHeight(36), 'medium'],
    small: ['small', calcLineHeight(28), 'small'],
    xsmall: ['xsmall', calcLineHeight(24), 'xsmall'],
  }
  const [fontSize, lineHeight, px] = sizes[props.size || 'medium']
  return {
    fontSize,
    lineHeight,
    px: !props.p ? px : undefined,
  }
}

function iconMargins(props: ThemedProps<ButtonProps>) {
  const spacing = { large: 0, small: 0 }
  switch (props.size) {
    case 'xsmall':
      spacing.small = 2
      spacing.large = 4
      break
    case 'small':
    case 'large':
    default:
      spacing.small = 4
      spacing.large = 8
  }

  if (props.iconBefore) {
    return css`
      margin-left: -${rem(spacing.small)};
      margin-right: ${rem(spacing.large)};
    `
  } else if (props.iconAfter) {
    return css`
      margin-left: ${rem(spacing.large)};
      margin-right: -${rem(spacing.small)};
    `
  } else {
    return false
  }
}

function getIcon(iconName: IconNames | undefined) {
  return iconName ? <ButtonIcon name={iconName} /> : null
}

// color is extracted here to ensure it is not passed to Box, creating a type
// error with the DOM's own color attribute.
const InternalButton: ComponentType = ({
  iconBefore,
  iconAfter,
  size,
  style,
  ...props
}) => {
  const { lineHeight, ...sizeHelperProps } = sizeHelper({
    size,
    ...props,
  })

  return (
    <Box
      is="button"
      borderRadius={CustomizableButtonAttributes.borderRadius}
      fontFamily="brand"
      py={props.p ? undefined : 'none'}
      display="inline-flex"
      alignItems="center"
      {...sizeHelperProps}
      {...omit(props, 'color')}
      style={{ lineHeight, ...style }}
    >
      {getIcon(iconBefore)}
      {props.children}
      {getIcon(iconAfter)}
    </Box>
  )
}

const ButtonFactory = React.forwardRef<
  StyledComponentType,
  ThemedProps<ButtonProps>
>((props: ThemedProps<ButtonProps>, ref: Ref<StyledComponentType>) => (
  <InternalButton innerRef={ref} {...props} />
))

export const CustomizableButtonAttributes: CustomizableAttributes = {
  borderRadius: 'medium',
}

const ButtonIcon = styled(Icon)``

/** @component */
export const Button = styled<ComponentType>(withTheme(ButtonFactory))`
  font-weight: 600;
  outline: none;
  transition: border 80ms;
  vertical-align: middle;
  white-space: nowrap;
  ${variantHelper};

  ${ButtonIcon} {
    ${iconMargins};
  }
`
