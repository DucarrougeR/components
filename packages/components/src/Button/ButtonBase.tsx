/*

 MIT License

 Copyright (c) 2019 Looker Data Sciences, Inc.
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  CompatibleHTMLProps,
  radii,
  RadiusSizes,
  reset,
  SemanticColors,
  space,
  SpaceProps,
} from 'looker-design-tokens'
import React, { forwardRef, Ref } from 'react'
import styled, { css } from 'styled-components'
import {
  minWidth,
  MinWidthProps,
  maxWidth,
  MaxWidthProps,
  WidthProps,
  width,
} from 'styled-system'
import { buttonSize, ButtonSizeProps } from './size'
import { ButtonIcon, buttonIcon, ButtonIconProps } from './icon'

export interface CustomizableButtonAttributes {
  borderRadius: RadiusSizes
}

export const CustomizableButtonAttributes: CustomizableButtonAttributes = {
  borderRadius: 'medium',
}

type ButtonColors = keyof SemanticColors

export interface ButtonProps
  extends Omit<CompatibleHTMLProps<HTMLButtonElement>, 'type'>,
    ButtonIconProps,
    ButtonSizeProps,
    MaxWidthProps,
    MinWidthProps,
    WidthProps,
    SpaceProps {
  type?: 'button' | 'submit' | 'reset'
  forwardedAs?: 'a'

  /**
   * Defines the color of the button. Can be the string name of a color listed in the color theme, or a color object.
   * @default "primary"
   */
  color?: ButtonColors

  className?: string
}

const ButtonJSX = forwardRef(
  (props: ButtonProps, ref: Ref<HTMLButtonElement>) => {
    const { children, className, iconBefore, iconAfter, ...restProps } = props

    return (
      <button className={className} ref={ref} {...restProps}>
        {iconBefore && <ButtonIcon name={iconBefore} />}
        {children}
        {iconAfter && <ButtonIcon name={iconAfter} />}
      </button>
    )
  }
)

ButtonJSX.displayName = 'ButtonJSX'

export const buttonCSS = css`
  ${reset}
  ${maxWidth}
  ${minWidth}
  ${width}

  align-items: center;
  border-radius: ${radii[CustomizableButtonAttributes.borderRadius]};
  cursor: pointer;
  display: inline-flex;
  font-weight: 600;
  outline: none;
  transition: border 80ms;
  vertical-align: middle;
  white-space: nowrap;

  &[disabled] {
    cursor: default;
    filter: grayscale(0.3);
    opacity: 0.25;
  }

  ${buttonSize}
  ${buttonIcon}

  ${space}
`

export const ButtonBase = styled(ButtonJSX)<ButtonProps>`
  ${buttonCSS}
`

ButtonBase.defaultProps = { size: 'medium' }
