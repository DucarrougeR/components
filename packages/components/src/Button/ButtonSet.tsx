import React, {
  ChangeEvent,
  forwardRef,
  ForwardRefExoticComponent,
  Ref,
} from 'react'
import {
  CompatibleHTMLProps,
  BorderProps,
  PositionProps,
  SpaceProps,
} from 'looker-design-tokens'
import { Box } from '../Layout'
import { ButtonItemProps } from './ButtonItem'

export interface ButtonGroupOrToggleProps<
  ValueType extends string | string[] = string[]
>
  extends PositionProps,
    BorderProps,
    SpaceProps,
    Omit<CompatibleHTMLProps<HTMLDivElement>, 'value' | 'onChange'> {
  /**
   * Internal use only
   */
  isControlled?: boolean
  /**
   * One or more ButtonItem
   */
  children: React.ReactNode
  /**
   * Value for controlling the component
   */
  value?: ValueType
  /**
   * Change callback for controlling the component
   */
  onChange?: (value: ValueType) => void
}

interface ButtonSetProps<ValueType extends string | string[] = string[]>
  extends Omit<ButtonGroupOrToggleProps<ValueType>, 'onChange'> {
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void
  isToggle?: boolean
}

export type ButtonSetType<
  T extends string | string[] = string[]
> = ForwardRefExoticComponent<ButtonSetProps<T> & { ref: Ref<HTMLDivElement> }>

export const ButtonSet = forwardRef(
  (
    {
      children,
      disabled,
      isToggle,
      name,
      onChange: groupOnChange,
      value,
      ...props
    }: ButtonSetProps,
    ref: Ref<HTMLDivElement>
  ) => {
    const clonedChildren = React.Children.map(children, child => {
      if (!React.isValidElement(child)) return child

      const { props: childProps } = child
      const childValue =
        childProps.value ||
        (typeof childProps.children === 'string' ? childProps.children : null)

      return React.cloneElement<ButtonItemProps>(child, {
        isControlled: groupOnChange !== undefined,
        name,
        type: isToggle ? 'radio' : 'checkbox',
        value: childValue,
        // pass down these optional props only if they're defined (overriding child props)
        ...(value && value.length !== 0
          ? {
              selected: isToggle
                ? value === childValue
                : value.indexOf(childValue) > -1,
            }
          : { selected: childProps.selected }),
        ...(disabled ? { disabled } : {}),
        ...(groupOnChange
          ? {
              onChange: (e: React.ChangeEvent<HTMLInputElement>) => {
                groupOnChange && groupOnChange(e)
                childProps.onChange && childProps.onChange(e)
              },
            }
          : {}),
      })
    })

    return (
      <Box
        alignItems="center"
        display="inline-flex"
        textAlign="center"
        fontSize="small"
        ref={ref}
        {...props}
      >
        {clonedChildren}
      </Box>
    )
  }
)

ButtonSet.displayName = 'ButtonSet'
