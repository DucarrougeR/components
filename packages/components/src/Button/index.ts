export * from './Button'
export * from './IconButton'
export * from './ButtonItem'
export * from './ButtonGroup'
export * from './ButtonOutline'
export * from './ButtonToggle'
export * from './ButtonTransparent'
