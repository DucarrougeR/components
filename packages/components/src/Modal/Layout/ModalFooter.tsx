import React, { FC } from 'react'
import styled from 'styled-components'
import { CompatibleHTMLProps } from 'looker-design-tokens'
import { SpaceProps, space } from 'styled-system'

export interface ModalFooterProps extends CompatibleHTMLProps<HTMLDivElement> {
  /**
   * Secondary content to place in the footer
   */
  secondary?: React.ReactNode
}

export const ModalFooter: FC<ModalFooterProps> = ({
  children,
  secondary,
  ...props
}) => {
  return (
    <Layout {...props}>
      <SpaceChildren>{children}</SpaceChildren>
      {secondary && <SpaceChildren mr="auto">{secondary}</SpaceChildren>}
    </Layout>
  )
}

const Layout = styled.footer`
  align-items: center;
  display: flex;
  flex-direction: row-reverse;
  padding: ${props => props.theme.space.large}
    calc(${p => p.theme.space.xlarge} - ${p => p.theme.space.xsmall});
  /**
    * Subtract margin placed on children from the horizontal padding applied to the component
    * so that the content is properly aligned.
    * padding is: vertical: large, horizontal: calc(xlarge - xsmall)
    **/
`

const SpaceChildren = styled.div<SpaceProps>`
  ${space};
  display: flex;
  flex-direction: row-reverse;

  & > * {
    margin-right: ${props => props.theme.space.xsmall};
    margin-left: ${props => props.theme.space.xsmall};
  }
`
