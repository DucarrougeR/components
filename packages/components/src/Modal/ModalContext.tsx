import { createContext, SyntheticEvent } from 'react'

export interface ModalContextProps {
  closeModal?: (event?: SyntheticEvent, doCallbacks?: boolean) => void
}

const context: ModalContextProps = {}

export const ModalContext = createContext(context)
