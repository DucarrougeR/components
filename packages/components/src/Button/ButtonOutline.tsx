import styled from 'styled-components'
import { ButtonBase } from './ButtonBase'

export const ButtonOutline = styled(ButtonBase)`
  background: ${({ theme, color = 'primary' }) =>
    theme.colors.semanticColors[color].white};
  border: 1px solid
    ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].borderColor};
  color: ${({ theme, color = 'primary' }) =>
    theme.colors.semanticColors[color].main};

  &:active,
  &.active {
    background: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].main};
    border-color: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].main};
    color: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].text};
  }

  &:hover,
  &:focus,
  &.hover {
    background: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].white};
    border-color: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].main};
    color: ${({ theme, color = 'primary' }) =>
      theme.colors.semanticColors[color].darker};
  }

  &[disabled] {
    &:hover,
    &:active,
    &:focus {
      background-color: ${({ theme, color = 'primary' }) =>
        theme.colors.semanticColors[color].white};
      border-color: ${({ theme, color = 'primary' }) =>
        theme.colors.semanticColors[color].borderColor};
      color: ${({ theme, color = 'primary' }) =>
        theme.colors.semanticColors[color].main};
    }
  }
`
