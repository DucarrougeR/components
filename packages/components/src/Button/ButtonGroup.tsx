import uniqueId from 'lodash/uniqueId'
import xor from 'lodash/xor'
import React, { ChangeEvent, FC, forwardRef, Ref } from 'react'
import styled from 'styled-components'
import { useControlWarn } from '../utils'
import { ButtonItem, ButtonItemLabel } from './ButtonItem'
import { ButtonGroupOrToggleProps, ButtonSet } from './ButtonSet'

const ButtonGroupFactory = forwardRef(
  (
    { onChange, value, ...props }: ButtonGroupOrToggleProps<string[]>,
    ref: Ref<HTMLDivElement>
  ) => {
    const isControlled = useControlWarn({
      controllingProps: ['onChange', 'value'],
      isControlledCheck: () => onChange !== undefined,
      name: 'ButtonGroup',
    })

    function handleChange(e: ChangeEvent<HTMLInputElement>) {
      const newValue = xor(value, [e.target.value])
      onChange && onChange(newValue)
    }

    return (
      <ButtonSet
        {...props}
        ref={ref}
        {...(isControlled
          ? {
              onChange: handleChange,
              value,
            }
          : { name: props.name || uniqueId() })}
      />
    )
  }
)

export const ButtonGroup = styled<FC<ButtonGroupOrToggleProps<string[]>>>(
  ButtonGroupFactory
)`
  ${ButtonItem} {
    height: 36px;
    border-radius: 4px;

    + ${ButtonItem} {
      margin-left: ${props => props.theme.space.xxsmall};
    }
  }

  ${ButtonItemLabel} {
    border-style: solid;
    border-width: 1px;
  }
`
