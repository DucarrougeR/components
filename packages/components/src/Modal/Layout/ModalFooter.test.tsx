import 'jest-styled-components'
import React from 'react'
import { assertSnapshot } from '@looker/components-test-utils'

import { Button } from '../../Button'
import { ModalContext } from '../ModalContext'
import { ModalFooter } from './ModalFooter'

test('ModalFooter with Button', () => {
  assertSnapshot(
    <ModalFooter>
      <Button variant="transparent">Cancel</Button>
    </ModalFooter>
  )
})

test('ModalFooter with ModalContext', () => {
  assertSnapshot(
    <ModalContext.Consumer>
      {({ closeModal }) => (
        <ModalFooter>
          <Button onClick={closeModal} variant="transparent">
            Cancel
          </Button>
          <Button
            onClick={() => {
              alert("doin' things...")
              close && close()
            }}
            color="danger"
          >
            Yes, Delete "Stuff"
          </Button>
        </ModalFooter>
      )}
    </ModalContext.Consumer>
  )
})
