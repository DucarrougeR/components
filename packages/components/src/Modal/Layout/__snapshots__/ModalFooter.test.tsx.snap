// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalFooter with Button 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
  padding: 1.25rem calc(2rem - 0.5rem);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c1 > * {
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

<footer
  className="c0"
>
  <div
    className="c1"
  >
    <button>
      Cancel
    </button>
  </div>
</footer>
`;

exports[`ModalFooter with ModalContext 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
  padding: 1.25rem calc(2rem - 0.5rem);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c1 > * {
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

<footer
  className="c0"
>
  <div
    className="c1"
  >
    <button>
      Cancel
    </button>
    <button
      onClick={[Function]}
    >
      Yes, Delete "Stuff"
    </button>
  </div>
</footer>
`;
